src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -389,9 +389,11 @@ private static unsafe int AlpnServerSelectCallback(IntPtr ssl, out byte* outp, o
             }
             catch
             {
+                protocolHandle.Free();
                 return Ssl.SSL_TLSEXT_ERR_NOACK;
```

Comment:
This doesn't work, it actually doesn't free the original handle, it just frees the copy, you should unskip the nonmatching alpn protocol test to verify it passes with the change.

Reply:
Its not how GCHandle works. While it's true the struct is copied, it is actually a slot (or in the code a cookie) in the GC with a ref to the object.

If you "free" a copy, it doesn't update the "allocated" state on the original struct, but it does actually mean the original will see "null" as it's target. Check the below example code

``` csharp
static void Main(string[] args)
{
        var handle = GCHandle.Alloc(new byte[100]);

        var ptr = GCHandle.ToIntPtr(handle);

        var handle2 = GCHandle.FromIntPtr(ptr);
        Console.WriteLine(handle.Target == null);
        handle2.Free();

        Console.WriteLine($"handle target is null {handle.Target == null}");
}
```

This shows that freeing the copy actually makes the original's "target" property == null.

Reply:
@stephentoub actually brought up the fact that I am not 100% sure on, that you can change the target and that works, in theory it should but I have never seen anyone "reallocate" the pinned object on the slot after it has been allocated. If it doesn't work, then it's potentially a bug, or a mistake to be allowed to "set" the target property.